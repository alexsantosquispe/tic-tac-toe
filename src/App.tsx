import { Board } from './components/Board';
import { Footer } from './components/Footer';
import { Navbar } from './components/Navbar';
import { ResetButton } from './components/ResetButton';
import useGame from './hooks/useGame';

function App() {
  const { winner, currentPlayer, resetGame, isDraw } = useGame();
  return (
    <>
      <div className="flex min-h-screen flex-col font-normal text-neutral-700 uppercase">
        <Navbar />
        <main className="flex flex-1 py-8 md:p-0">
          <section className="flex w-full flex-col items-center justify-center gap-6 px-4 md:gap-10">
            <h1 className="flex flex-col gap-2 text-center">
              <span className="text-5xl font-medium">
                TIC-<span className="text-rose-600">TAC</span>-TOE
              </span>
              <span className="text-xl md:text-2xl">
                Lets play a classic game!
              </span>
            </h1>
            <div className="flex flex-col items-center justify-between gap-4 md:gap-8">
              <p className="text-3xl md:text-4xl" aria-live="polite">
                {!winner && !isDraw && currentPlayer && `${currentPlayer} turn`}
                {winner && `${winner} wins!`}
                {isDraw && !winner && 'Draw!'}
              </p>
              <Board />
              <ResetButton onClick={resetGame} />
            </div>
          </section>
        </main>
        <Footer />
      </div>
    </>
  );
}

export default App;
