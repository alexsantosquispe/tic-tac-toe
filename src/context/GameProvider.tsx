import { useState, type ReactNode } from 'react';
import { DEFAULT_DATA } from '../constants';
import type { SquareValueTypes } from '../types';
import { getWinnerResult } from '../utils';
import GameContext from './GameContext';

interface GameProviderProps {
  children: ReactNode;
}

const GameProvider = ({ children }: GameProviderProps) => {
  const [currentPlayer, setCurrentPlayer] = useState<'X' | 'O'>('X');
  const [data, setData] = useState<SquareValueTypes[]>(DEFAULT_DATA);
  const [winner, setWinner] = useState<SquareValueTypes | null>(null);
  const [combHighLight, setCombHighLight] = useState<number[]>([]);

  const toggleCurrentPlayer = () => {
    setCurrentPlayer((prevPlayer) => (prevPlayer === 'X' ? 'O' : 'X'));
  };

  const checkSquare = (index: number) => {
    if (data[index] !== '' || winner) return;
    const newData = [...data];
    newData[index] = currentPlayer;
    setData(newData);
    const result = getWinnerResult(newData);
    if (!result.winner) {
      toggleCurrentPlayer();
    }
    setWinner(result.winner);
    setCombHighLight(result.combHighLight);
  };

  const resetGame = () => {
    setData(DEFAULT_DATA);
    setWinner(null);
    setCombHighLight([]);
    setCurrentPlayer('X');
  };

  return (
    <GameContext.Provider
      value={{
        data,
        currentPlayer,
        winner,
        combHighLight,
        checkSquare,
        toggleCurrentPlayer,
        resetGame
      }}
    >
      {children}
    </GameContext.Provider>
  );
};

export default GameProvider;
