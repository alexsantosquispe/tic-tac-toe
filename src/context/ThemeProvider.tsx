import { useEffect, useState, type ReactNode } from 'react';
import {
  applyTheme,
  getStoredTheme,
  getSystemTheme,
  setStoredTheme
} from '../utils/theme.utils';
import { THEME_TYPES, ThemeContext, type ThemeType } from './ThemeContext';

interface ThemeProviderProps {
  children: ReactNode;
}

const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const initialTheme = getStoredTheme();
  const [currentTheme, setCurrentTheme] = useState<ThemeType>(
    initialTheme ?? THEME_TYPES.SYSTEM
  );

  useEffect(() => {
    const theme =
      currentTheme === THEME_TYPES.SYSTEM ? getSystemTheme() : currentTheme;
    applyTheme(theme);
    setStoredTheme(currentTheme);
  }, [currentTheme]);

  useEffect(() => {
    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    const handleChange = (e: MediaQueryListEvent) => {
      if (getStoredTheme() === THEME_TYPES.SYSTEM) {
        applyTheme(e.matches ? THEME_TYPES.DARK : THEME_TYPES.LIGHT);
      }
    };
    mediaQuery.addEventListener('change', handleChange);
    return () => mediaQuery.removeEventListener('change', handleChange);
  }, []);

  return (
    <ThemeContext.Provider
      value={{ theme: currentTheme, setTheme: setCurrentTheme }}
    >
      {children}
    </ThemeContext.Provider>
  );
};

export default ThemeProvider;
