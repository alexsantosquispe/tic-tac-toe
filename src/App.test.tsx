import '@testing-library/jest-dom';

import {
  fireEvent,
  render,
  screen,
  waitFor,
  within
} from '@testing-library/react';

import App from './App';
import GameProvider from './context/GameProvider';
import ThemeProvider from './context/ThemeProvider';
import { I18NWrapper } from './tests/testsUtils';

describe('App', () => {
  beforeEach(async () => {
    await waitFor(() => {
      render(
        <I18NWrapper>
          <ThemeProvider>
            <GameProvider>
              <App />
            </GameProvider>
          </ThemeProvider>
        </I18NWrapper>
      );
    });
  });

  describe('styles', () => {
    it('should render the app correctly', async () => {
      await waitFor(() => {
        expect(screen.getByText('X turn')).toBeInTheDocument();
        expect(screen.getByTestId('board')).toBeInTheDocument();
        expect(
          screen.getByRole('button', { name: 'Reset game button' })
        ).toBeInTheDocument();
      });
    });
  });

  describe('behavior', () => {
    it('should change the current player after select a square', () => {
      const board = screen.getByTestId('board');

      const squareButtons = within(board).getAllByRole('button');

      expect(screen.getByText('X turn')).toBeInTheDocument();

      fireEvent.click(squareButtons[0]);

      expect(screen.getByText('O turn')).toBeInTheDocument();
    });

    it('should show the winner', () => {
      const board = screen.getByTestId('board');
      const squareButtons = within(board).getAllByRole('button');

      fireEvent.click(squareButtons[0]); //X
      fireEvent.click(squareButtons[3]); //O
      fireEvent.click(squareButtons[1]); //X
      fireEvent.click(squareButtons[4]); //O
      fireEvent.click(squareButtons[2]); //X

      expect(screen.getByText('X wins!')).toBeInTheDocument();
    });

    it('should check draw', () => {
      const board = screen.getByTestId('board');
      const squareButtons = within(board).getAllByRole('button');

      fireEvent.click(squareButtons[0]); //X
      fireEvent.click(squareButtons[1]); //O
      fireEvent.click(squareButtons[2]); //X
      fireEvent.click(squareButtons[3]); //O
      fireEvent.click(squareButtons[4]); //X
      fireEvent.click(squareButtons[6]); //0
      fireEvent.click(squareButtons[5]); //X
      fireEvent.click(squareButtons[8]); //O
      fireEvent.click(squareButtons[7]); //X

      expect(screen.getByText('Draw!')).toBeInTheDocument();
    });

    it('should reset the game', () => {
      const board = screen.getByTestId('board');
      const squareButtons = within(board).getAllByRole('button');

      const resetButton = screen.getByRole('button', {
        name: 'Reset game button'
      });

      fireEvent.click(squareButtons[0]);
      fireEvent.click(squareButtons[1]);

      expect(within(board).getAllByTestId('x-icon')[0]).toBeInTheDocument();
      expect(
        within(board).getAllByTestId('circle-icon')[0]
      ).toBeInTheDocument();

      fireEvent.click(resetButton);

      waitFor(() => {
        expect(within(board).getAllByTestId('x-icon').length).toBe(0);
        expect(within(board).getAllByTestId('circle-icon').length).toBe(0);
      });
    });

    it('should open and close de settings modal', () => {
      const settingsButton = screen.getByRole('button', {
        name: 'Settings button'
      });

      expect(settingsButton).toBeInTheDocument();

      fireEvent.click(settingsButton);

      const settingsModal = screen.getByTestId('modal');

      expect(settingsModal).toBeInTheDocument();

      const closeSettingsButton = screen.getByRole('button', {
        name: 'Close modal button'
      });

      fireEvent.click(closeSettingsButton);

      waitFor(() => {
        expect(settingsModal).not.toBeInTheDocument();
      });
    });
  });
});
