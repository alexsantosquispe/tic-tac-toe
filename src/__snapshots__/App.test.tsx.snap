// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`App styles should render the app correctly 1`] = `undefined`;

exports[`App styles should render the app correctly 2`] = `undefined`;

exports[`App styles should render the app correctly 3`] = `undefined`;
